File: Exámenes/2016 FEB/prEulerianCycle/src/dataStructures/graph/EulerianCycle.java
/**
 * Student's name: Ángel Manuel Soria Gil
 * Student's group: Doble Grado Matemáticas e Ingeniería Informática
 * Data Structures. E.T.S.I. Informática. UMA.
 */

package dataStructures.graph;
import dataStructures.dictionary.Dictionary;
import dataStructures.dictionary.HashDictionary;
import dataStructures.set.*;
import dataStructures.list.*;

public class EulerianCycle<V> {
    private List<V> eCycle;

    @SuppressWarnings("unchecked")
    public EulerianCycle(Graph<V> g) {
        Graph<V> graph = (Graph<V>) g.clone();
        eCycle = eulerianCycle(graph);
    }

    public boolean isEulerian() {
        return eCycle != null;
    }

    public List<V> eulerianCycle() {
        return eCycle;
    }

    // J.1
    private static <V> boolean isEulerian(Graph<V> g) {
        Boolean res = true;
        for (V v : g.vertices()){
            if (g.degree(v)%2!=0){
                res = false;
            }
        }
        return res;
    }

    // J.2
    private static <V> void remove(Graph<V> g, V v, V u) {
        g.deleteEdge(v, u);
        if (g.degree(v)==0){
            g.deleteVertex(v);
        }
        if (g.degree(u)==0){
            g.deleteVertex(u);
        }
    }

    // J.3
    private static <V> List<V> extractCycle(Graph<V> g, V v0) {
        List<V> res = new LinkedList<>();
        res.append(v0);
        V actual = v0;
        V next = g.successors(v0).iterator().next();
        while (!next.equals(actual)){
            res.append(next);
            remove(g, actual, next);
            actual = next;
            next = g.successors(actual).iterator().next();
        }
        return res;
    }

    // J.4
    private static <V> void connectCycles(List<V> xs, List<V> ys) {
        if (xs.isEmpty()){
            xs = ys;
        } else {
            List <V> res = new LinkedList<>();
            int i = 0;
            V elem = ys.get(i);
            while (!xs.get(i).equals(elem)){
                res.append(xs.get(i));
                i++;
            }
            for (V vertex: ys){
                res.append(vertex);
            }
            for (i=i+1; i<xs.size(); i++){
                res.append(xs.get(i));
            }
            xs = res;
        }

    }

    // J.5
    private static <V> V vertexInCommon(Graph<V> g, List<V> cycle) {
        V res = null;
        Set<V> vertex = g.vertices();
        for (V v : cycle){
            while (vertex.iterator().hasNext()){
                if (vertex.iterator().next().equals(v)){
                    res = v;
                    break;
                }
            }
            break;
        }
        return res;
    }

    // J.6
    private static <V> List<V> eulerianCycle(Graph<V> g) {
        List<V> res = new LinkedList<>();
        if (isEulerian(g)){
            V v = g.vertices().iterator().next();
            res = extractCycle(g,v);
            while (!g.isEmpty()){
                V u = vertexInCommon(g, res);
                List<V> c2 = extractCycle(g, u);
                connectCycles(res, c2);
            }
        } else {
            //throw new GraphException("error: applying eulerianCycle on not eulerian graph");
            res = null;
        }
        return res;
    }
}

File: Exámenes/2016 FEB/eulerianCycle2/haskell/DataStructures/Graph/EulerianCycle.hs
-------------------------------------------------------------------------------
-- Student's name: Ángel Manuel Soria Gil
-- Student's group: Doble Grado Matemáticas e Ingeniería Informática D
--
-- Data Structures. Grado en Informática. UMA.
-------------------------------------------------------------------------------

module DataStructures.Graph.EulerianCycle(isEulerian, eulerianCycle) where

import DataStructures.Graph.Graph
import Data.List

--H.1)
isEulerian :: Eq a => Graph a -> Bool
isEulerian g = isEulerianAux (vertices g)
    where 
        isEulerianAux (v:rest) = even (degree g v) && isEulerianAux rest
        isEulerianAux []       = True 

-- H.2)
remove :: (Eq a) => Graph a -> (a,a) -> Graph a
remove g (v,u)
    | degree g v <=1 && degree g u <=1 = deleteVertex (deleteVertex (deleteEdge g (v,u)) v) u
    | degree g v <= 1                  = deleteVertex (deleteEdge g (v,u)) v 
    | degree g u <= 1                  = deleteVertex (deleteEdge g (v,u)) u 
    | otherwise                        = deleteEdge g (v,u)

-- H.3)
extractCycle :: (Eq a) => Graph a -> a -> (Graph a, Path a)
extractCycle g v0 = extractCycleAux g v0 v0 (head (successors g v0)) (g, [v0])
    where
        extractCycleAux g v0 v u (g', cycle) 
            | v0 == u   = ((remove g' (v,u)), cycle++[u])
            | otherwise = extractCycleAux g' v0 u (head (successors g' u)) ((remove g' (v,u)), cycle++[u]) 

-- H.4)
connectCycles :: (Eq a) => Path a -> Path a -> Path a
connectCycles [] ys      = ys
connectCycles xs (y:ys)  = xs1 ++ init (y:ys) ++ xs2
    where 
        f elem = y == elem
        xs1    = takeWhile (not.f) xs
        xs2    = dropWhile (not.f) xs 
        
lista1 = ['a', 'c', 'd', 'a']
lista2 = ['c', 'b', 'e', 'c']

-- H.5)
vertexInCommon :: Eq a => Graph a -> Path a -> a
vertexInCommon g cycle = head [v | v <- vertices g, v `elem` cycle]

-- H.6) 
eulerianCycle :: Eq a => Graph a -> Path a
eulerianCycle g 
    | isEmpty g    = []
    | isEulerian g = eulerianCycleAux g (head (vertices g)) []
    | otherwise    = error "applied eulerianCycle on not eulerianCycle"

eulerianCycleAux :: Eq a => Graph a -> a -> Path a -> Path a 
eulerianCycleAux g v path
    | isEmpty g = path
    | otherwise = eulerianCycleAux g' v' path'
    where
        path'   = (connectCycles path p)
        (g', p) = extractCycle g v 
        v'      = vertexInCommon g' path'

-----------------------------------------------------------------------------
-----------------------------------------------------------------------------------------
data Vertex = A | B | C | D | E  deriving (Show,Eq,Enum,Ord)

g5 :: Graph Vertex
g5 = mkGraphEdges [A .. E]
                  [(A, C), (A, D), (B, C), (B, E), (C, D), (C, E)]

g3 :: Graph Vertex
g3 = fst (extractCycle g5 A)